spices/SPICE-0010-overhauled-mapping-listing-typechecks.adoc

```diff
@@ -0,0 +1,432 @@
+= Overhauled Mapping/Listing Typechecks
+
+* Proposal: link:./SPICE-0010-overhauled-mapping-listing-typechecks.adoc[SPICE-0010]
+* Status: TBD
+* Implemented in: Pkl TBD
+* Category: Language
+
+== Introduction
+
+During typechecks for Listings and Mappings, retain laziness of their object members.
+
+== Motivation
+
+Currently, in order to typecheck a listing or mapping, Pkl will shallow-force its members.
+
+For example, in the following snippet:
+
+[source,pkl]
+----
+class Bird { name: String }
+
+hidden birds: Listing<Bird> = new {
+  new { name = "Osprey" }
+  new { name = "Warbler" }
+}
+
+numBirds = birds[0]
+//         ^^^^^
+----
+
+During execution of `birds` (as highlighted by the carets), Pkl will execute both members of `birds`, including `new { name = "Warbler" }`, even though it is not used in producing the program's result.
+
+In psuedocode, Pkl executes the following instructions:
+
+[source]
+----
+assert value instanceof Listing
+for elem in value.members():
+  assert elem.eval() instanceof Bird
+----
+
+There are two underlying issues with this approach.
+
+=== Performance cost
```

Comment:
Note that the current implementation of surrogate objects is itself quite expensive: It adds 5 fields to, and allocates 2 empty collections for each instance of, class `VmListing` and `VmMapping` (via a common superclass). These classes aren't only used for surrogate objects but also for regular listings/mappings. Additionally, case distinctions for surrogate objects are added throughout the codebase.

One way to improve on the current implementation is to move towards an object model that supports different internal representations for the same type of Pkl object. For example, different representations could be used for small, large, and surrogate listings/mappings.